Motorola Surnia Backlight Driver V2

BogdanTheGeek opened the pull request:
Added support for brightness control for motorola-surnia (sprey and styx too) though led-backlight driver and rohm bd65b60 driver.

Re-submission of #284

I re-wrote the driver from scratch, it was more fun than I expected. Hopefully this time its easier to review. I've done my best to check the coding standards, examples and old PRs for guidance.

BogdanTheGeek committed:
leds: Add support for rohm,bd65b60 led driver

The chip supports 2 outpus sharing the same current setting
and is controlled over I2C.

Signed-off-by: Bogdan Ionescu <bogdan.ionescu.work@gmail.com>

BogdanTheGeek committed:
arm64: dts: qcom: msm8916-motorola-surnia: Add backlight control

Enabling backlight brightness control for motorola-surnia.

Signed-off-by: Bogdan Ionescu <bogdan.ionescu.work@gmail.com>

BogdanTheGeek committed:
dt-bindings: leds: Document rohm,bd65b60

Added documentation for the rohm,bd65b60 led driver.

Signed-off-by: Bogdan Ionescu <bogdan.ionescu.work@gmail.com>

TravMurav commented on drivers/leds/leds-bd65b60.c:
```diff
@@ -0,0 +1,282 @@
+// SPDX-License-Identifier: GPL-2.0+
```
Is the original driver "+"?
Did you intend to use the "-or-later"?

BogdanTheGeek replied:
https://www.kernel.org/doc/html/v5.0/process/license-rules.html
GPL-2.0+  :  GNU General Public License v2.0 or later
I assumed this was the new format and meant the same thing.

TravMurav replied:
SPDX spec deprecates the old notation with + and prefers -only and -or-later suffixes instead

https://spdx.org/licenses/ - bottom has a list of deprecated identifiers

I will assume that the original driver allowed -or-later explicitly and you want to keep that

TravMurav commented on drivers/leds/leds-bd65b60.c:
```diff
@@ -0,0 +1,282 @@
+
+	/* Check required properties */
+	if (!fwnode_property_present(child, "rohm,enable")) {
+		dev_err(dev, "No reg property found");
```
Doesn't look like reg :)

TravMurav commented on drivers/leds/leds-bd65b60.c:
```diff
@@ -0,0 +1,282 @@
+		return -ENOENT;
+	}
+
+	ret = fwnode_property_read_u32(child, "rohm,enable", &led->enable);
+	if (ret || (led->enable & LEDSEL_MASK) != led->enable) {
+		dev_err(dev, "Failed to read enable property");
+		return ret;
+	}
```
The property name seems rather indescriptive... This seems to define what led strings are enabled, so maybe enabled-outputs or something?

TravMurav commented on drivers/leds/leds-bd65b60.c:
```diff
@@ -0,0 +1,282 @@
+		return ret;
+	}
+
+	mutex_init(&led->lock);
```
I think you're supposed to consider mutex a resource and explicitly destroy it.

TravMurav commented on drivers/leds/leds-bd65b60.c:
```diff
@@ -0,0 +1,282 @@
+	struct bd65b60_led *led = i2c_get_clientdata(client);
+
+	if (regmap_write(led->regmap, REG_PON, BD65B60_OFF))
+		dev_err(&client->dev, "Failed to turn off led");
```
Would be nice to log the ret here too.

TravMurav commented on drivers/leds/leds-bd65b60.c:
```diff
@@ -0,0 +1,282 @@
+		return ret;
+	}
+
+	dev_info(&client->dev, "BD65B60 LED driver initialized");
```
This line doesn't really add much value IMO, let's not spam the dmesg. You can probably make it dev_dbg() if you want to keep it

TravMurav commented on drivers/leds/leds-bd65b60.c:
```diff
@@ -0,0 +1,282 @@
+module_i2c_driver(bd65b60_i2c_driver);
+
+MODULE_DESCRIPTION("ROHM Semiconductor led driver for bd65b60");
+MODULE_LICENSE("GPL");
```
Maybe could add MODULE_AUTHOR too?

TravMurav commented on include/dt-bindings/leds/leds-bd65b60.h:
```diff
@@ -0,0 +1,22 @@
+/* SPDX-License-Identifier: GPL-2.0 */
```
header change does not belong to a dts commit.

BogdanTheGeek replied:
Even though its a dts header?

TravMurav replied:
yes, this is a binding change that has nothing to do with the device specifically. Please move this into the binding commit instead.

TravMurav commented on include/dt-bindings/leds/leds-bd65b60.h:
```diff
@@ -0,0 +1,22 @@
+/* SPDX-License-Identifier: GPL-2.0 */
```
And this one is -only...

BogdanTheGeek replied:
I am the sole author for this header file.
Most of the other headers don't even have a licence, but the one that does has:
`GPL-2.0-only OR BSD-2-Clause`
Which one should I use?

TravMurav replied:
I'd prefer either `(GPL-2.0-only OR BSD-2-Clause)` or just `GPL-2.0-only` given this is how most of the binding headers are licensed... (for dts there is a recent push to have them as BSD-2-Clause but not sure if this applies to the binding docs given how many ones were made prior to that decision...)

TravMurav commented on Documentation/devicetree/bindings/leds/rohm,bd65b60-leds.yaml:
```diff
@@ -0,0 +1,68 @@
+    const: rohm,bd65b60
+
+patternProperties:
+  "^led-[1-2]$":
```
can it have two nodes? Maybe allow the pattern to match just `led` too?

TravMurav commented on Documentation/devicetree/bindings/leds/rohm,bd65b60-leds.yaml:
```diff
@@ -0,0 +1,68 @@
+        default: 16
+
+required:
+  - compatible
```
Can it work without reg? :)

TravMurav commented on Documentation/devicetree/bindings/leds/rohm,bd65b60-leds.yaml:
```diff
@@ -0,0 +1,68 @@
+  - rohm,enable
+additionalProperties: false
+
+examples:
```
Did you run the dt_binding_check on this file?

BogdanTheGeek replied:
did now :)

TravMurav commented on Documentation/devicetree/bindings/leds/rohm,bd65b60-leds.yaml:
```diff
@@ -0,0 +1,68 @@
+
+required:
+  - compatible
+  - rohm,enable
```
I think this requirement would go into the subnode definition...

TravMurav reviewed (commented):
Mostly nitpicks

TravMurav commented on include/dt-bindings/leds/leds-bd65b60.h:
```diff
@@ -0,0 +1,22 @@
+/* SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause) */
+/*
+ * This header provides macros for the ROHM BD65B60 device tre§e bindings.
```
typo in tre§e?

BogdanTheGeek replied:
vim + mac signature :) § is next to <Esc>

TravMurav commented on include/dt-bindings/leds/leds-bd65b60.h:
```diff
@@ -0,0 +1,22 @@
+#ifndef _DT_BINDINGS_LEDS_BD65B60_H
+#define _DT_BINDINGS_LEDS_BD65B60_H
+
+#define BIT(X) (1 << (X))
```
not sure if you may want to undefine this later or guard somehow if some other binding is included that defines bit too... Need to grep for headers maybe to see how others do it

BogdanTheGeek replied:
I'll just remove it

TravMurav commented on Documentation/devicetree/bindings/leds/rohm,bd65b60-leds.yaml:
```diff
@@ -0,0 +1,77 @@
+    #include <dt-bindings/leds/leds-bd65b60.h>
+
+    i2c {
+
```
this newline looks redundant

BogdanTheGeek replied:
keeping it in theme with the common.yaml

TravMurav reviewed (commented):
also the usual commit order is bindings then driver then dt changes adding users

BogdanTheGeek commented:
Re-ordered the commits, but I assume that only the first 2 will be upstreamed and the dts will go into [pmaports](https://gitlab.com/postmarketOS/pmaports/-/raw/master/device/community/linux-postmarketos-qcom-msm8916/config-postmarketos-qcom-msm8916.aarch64).

TravMurav commented on drivers/leds/leds-bd65b60.c:
```diff
@@ -0,0 +1,287 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
```
Commit message: Could be better written focusing on the led driver itself, it's not like it can be found \_only\_ in Motorola phones :)

TravMurav commented on Documentation/devicetree/bindings/leds/rohm,bd65b60-leds.yaml:
```diff
@@ -0,0 +1,77 @@
+# SPDX-License-Identifier: GPL-2.0-only OR BSD-2-Clause
```
Commit subject should be `dt-bindings: leds: Add/Document...` as all other files around it.

TravMurav commented on arch/arm64/boot/dts/qcom/msm8916-motorola-surnia.dts:
```diff
@@ -57,6 +59,31 @@
 		pinctrl-1 = <&usb_id_sleep>;
 		pinctrl-names = "default", "sleep";
 	};
+
```
Commit subject: `arm64: dts: qcom: msm8916-motorola-surnia: Add ...` like in the history of the file

TravMurav commented on arch/arm64/boot/dts/qcom/msm8916-motorola-surnia.dts:
```diff
@@ -57,6 +59,31 @@
 		pinctrl-names = "default", "sleep";
 	};
+
+	backlight_dsi: backlight-dsi {
```
```suggestion
	backlight: backlight {
```

TravMurav commented on arch/arm64/boot/dts/qcom/msm8916-motorola-surnia.dts:
```diff
@@ -57,6 +59,31 @@
+&blsp_i2c1 {
+	status = "okay";
+
+	led_controler@64 {
```
```suggestion
	led-controler@64 {
```

TravMurav commented on arch/arm64/boot/dts/qcom/msm8916-motorola-surnia.dts:
```diff
@@ -57,6 +59,31 @@
+	led_controler@64 {
+		compatible = "rohm,bd65b60";
+		reg = <0x64>;
+		pinctrl-names = "default";
+		pinctrl-0 = <&backlight_reset_default>;
```
```suggestion
		pinctrl-0 = <&backlight_reset_default>;
		pinctrl-names = "default";
```

TravMurav reviewed (commented):
Driver commit is still before the dt-bindings, please swap them around

BogdanTheGeek committed:
arm64: dts: qcom: msm8916-motorola-osprey: Add backlight control

Based on motorola-surnia, the only changes being:
 - Using LED2 instead of LED1 for the led driver
 - Using GPIO 113 for the backlight reset

Signed-off-by: Bogdan Ionescu <bogdan.ionescu.work@gmail.com>

TravMurav commented on arch/arm64/boot/dts/qcom/msm8916-motorola-surnia.dts:
```diff
@@ -57,6 +59,31 @@
 		pinctrl-names = "default", "sleep";
 	};
+
+	backlight: backlight{
```
```suggestion
	backlight: backlight {
```

BogdanTheGeek commented:
I'm not upstreaming code anymore, anyone can take this as written consent to upstream it as their own without any need to credit me if you want it in mainline.
